cpp/src/arrow/array.h

```diff
@@ -734,6 +745,27 @@ class ARROW_EXPORT DictionaryArray : public Array {
  private:
   void SetData(const std::shared_ptr<ArrayData>& data);
 
+  /// \brief Check if all indices are within valid range
+  ///
+  /// \param[in] indices dictionary indices
+  /// \param[in] range valid range of indices (0 <= index < range)
+  template <typename ArrowType>
+    static bool SanityCheck(const std::shared_ptr<Array>& indices, const int64_t range) {
+    using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+    std::shared_ptr<ArrayType> array = std::static_pointer_cast<ArrayType>(indices);
+    const typename ArrowType::c_type* data = array->raw_values();
+    const int64_t size = sizeof(data) / sizeof(data[0]);
```

Comment:
Use `array->length()` instead of this

Reply:
@xuepanchen Note that `sizeof` is a compile time construct, so it cannot reliably return the number of valid elements in a chunk of dynamically allocated memory. This trick only works on statically allocated C arrays because the number of bytes in them is known at compile time.